File: 序号/707.设计链表.kt
//执行用时：244 ms, 在所有 Kotlin 提交中击败了76.47% 的用户
// 内存消耗：40.5 MB, 在所有 Kotlin 提交中击败了82.35% 的用户
// 通过测试用例：64 / 64
class MyLinkedList() {
  data class Node(val i:Int,var next:Node?,var prev:Node?)
  var head:Node?=null
  var tail:Node?=null
  var size = 0
  private fun getNode(index:Int):Node?{
    if (size==0) return null
    if (index>=size) return null
    var p = head
    if (index<=size/2){
      for (i in 0 until index){
        p=p?.next
      }
      return p
    }else{
      var t =tail
      for (i in 0 until size-1-index){
        t = t?.prev
      }
      return t
    }
  }
  fun get(index: Int): Int {
    val node = getNode(index)?:return -1
    return node.i
  }
  
  fun addAtHead(`val`: Int) {
    val newHead = Node(`val`,null,null)
    if(size==0){
      head=newHead
      tail = head
    }else {
      newHead.next=head
      head?.prev = newHead
      head = newHead
    }
    size++
    
  }
  
  fun addAtTail(`val`: Int) {
    val newTail = Node(`val`,null,null)
    if(size==0){
      tail = newTail
      head = tail
    }else{
      tail?.next = newTail
      newTail.prev=tail
      tail = newTail
    }
    size++
  }
  
  
  fun addAtIndex(index: Int, `val`: Int) {
    if(index>size) return
    if (index<=0) addAtHead(`val`)
    else if(index==size) addAtTail(`val`)
    else{
      val node = getNode(index)
      val prev = node!!.prev
      val newNode=Node(`val`,node,prev)
      node?.prev?.next=newNode
      node?.prev=newNode
      size++
    }
  }
  
  fun deleteAtIndex(index: Int) {
    val node = getNode(index)?:return
    if(size==1){
      head=null
      tail = null
    }else if(node == head){
      head = head!!.next
    }else if(node == tail){
      tail = tail!!.prev
    }else {
      val prev = node.prev
      val next = node.next
      prev?.next = next
      next?.prev = prev
    }
    size--
  }
}

File: 序号/707.设计链表.cpp
//64 / 64 个通过测试用例
//   执行用时: 36 ms
//   内存消耗: 19.3 MB
class MyLinkedList {
  struct Node {
    int i;
    Node* next;
    Node* prev;
  };
  int size = 0;
  Node* head = nullptr;
  Node* tail = nullptr;

 private:
  Node* getNode(int index) {
    if (size == 0) return nullptr;
    if (index >= size) return nullptr;

    if (index <= size / 2) {
      Node* p = head;
      for (int i = 0; i < index; i++) {
        p = p->next;
      }
      return p;
    } else {
      Node* t = tail;
      for (int i = 0; i < size - 1 - index; i++) {
        t = t->prev;
      }
      return t;
    }
  }

 public:
  MyLinkedList() {}

  int get(int index) {
    Node* node = getNode(index);
    if (node) {
      return node->i;
    } else {
      return -1;
    }
  }

  void addAtHead(int val) {
    Node* newHead = new Node{val};
    if (size == 0) {
      head = newHead;
      tail = head;
    } else {
      newHead->next = head;
      head->prev = newHead;
      head = newHead;
    }
    size++;
  }

  void addAtTail(int val) {
    Node* newTail = new Node{val};
    if (size == 0) {
      tail = newTail;
      head = tail;
    } else {
      tail->next = newTail;
      newTail->prev = tail;
      tail = newTail;
    }
    size++;
  }

  void addAtIndex(int index, int val) {
    if(index>size) return;
    if (index <= 0) {
      addAtHead(val);
    } else if (index == size) {
      addAtTail(val);
    } else {
      Node* node = getNode(index);
      Node* prev = node->prev;
      Node* newNode = new Node{val};
      if (prev) {
        prev->next = newNode;
      }
      newNode->next = node;
      node->prev = newNode;
      newNode->prev = prev;
      size++;
    }
  }

  void deleteAtIndex(int index) {
    Node* node = getNode(index);
    if (!node) return;
    if (size == 1) {
      head = nullptr;
      tail = nullptr;
    } else if (node == head) {
      head = head->next;
    } else if (node == tail) {
      tail = tail->prev;
    } else {
      Node* prev = node->prev;
      Node* next = node->next;
      prev->next = next;
      next->prev = prev;
    }
    size--;
  }
};